lib/tokens/TokenManager.js

```diff
@@ -0,0 +1,117 @@
+// Copyright © 2019 IBM Corp. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+'use strict';
+
+const debug = require('debug')('cloudant:tokens:tokenmanager');
+const cookie = require('cookie');
+const EventEmitter = require('events');
+
+class TokenManager {
+  constructor(client, jar, sessionUrl) {
+    this._client = client;
+    this._jar = jar;
+    this._sessionUrl = sessionUrl;
+
+    this._forceRenew = true;
+    this._lastRenewResponse = null;
+    this._renewingToken = false;
+
+    this._ee = new EventEmitter().setMaxListeners(Infinity);
+  }
+
+  _autoRenew() {
+    debug('Auto renewing token now...');
+    this.forceRenew().then((response) => {
+      let maxAgeSecs = cookie.parse(response.headers['set-cookie'][0])['Max-Age'] || 3600000;
+      let delaySecs = maxAgeSecs / 2;
+      debug(`Renewing token in ${delaySecs} seconds.`);
+      setTimeout(this._autoRenew.bind(this), delaySecs * 1000).unref();
+    }).catch((error) => {
+      debug(`Failed to auto renew token - ${error}. Retrying in 60 seconds.`);
+      setTimeout(this._autoRenew.bind(this), 60000).unref();
+    });
+  }
+
+  _getToken(callback) {
+    // ** Method to be implemented by _all_ subclasses of `TokenManager` **
+    throw new Error('Not implemented.');
+  }
+
+  // Getter for `forceRenew`.
+  // - `true`  A renewal attempt will be made on the next `renew` request.
+  // - `false` No renewal attempt will be made on the next `renew`
+  //           request. Instead the last good renewal response will be returned
+  //           to the client.
+  get getforceRenew() {
```

Comment:
If we agree on renaming `_forceRenew` variable above, this function should also be renamed.